Add kernel image filters

fbudin69500 opened the pull request.

fbudin69500 committed:
ENH: Support of automatically generated KernelImage filters

Python script has been update to generate filters that require a kernel
as their 3rd template parameter. Supported kernels are:
* Annulus
* Ball
* Box
* Cross

fbudin69500 committed:
ENH: Addition of ITKBlackTopHatImage filter

fbudin69500 committed:
ENH: BlackTopHat Image filter uses an isotropic kernel

Kernel radius is set with one integer value instead of a vector.

thewtex commented on Utilities/SimpleITKJSONDream3DFilterCreationTemplates/KernelImageFilterTemplate.cpp:
```diff
@@ -0,0 +1,215 @@
+  sitkPolygon6,
+  sitkPolygon7,
+  sitkPolygon8,
+  sitkPolygon9
```
Remove Polygon members?

fbudin69500 replied:
Makes sense.

thewtex commented on Utilities/SimpleITKJSONDream3DFilterCreationTemplates/KernelImageFilterTemplateTest.cpp:
```diff
@@ -0,0 +1,66 @@
+  sitkPolygon6,
+  sitkPolygon7,
+  sitkPolygon8,
+  sitkPolygon9
```
Remove Polygon members?

thewtex commented on Documentation/ITKImageProcessingFilters/ITKBlackTopHatImage.md:
```diff
@@ -0,0 +1,49 @@
+## Description ##
+Black top hat extract local minima that are smaller than the structuring element.
+
+Black top hat extract local minima that are smaller than the structuring element. It subtract the background in the input image. The output of the filter transforms the black peaks in white peaks.
```
subtract -> subtracts

The output of the filter returns black structures that are smaller than the structuring element.

Please also fix these in the ITK documentation.

fbudin69500 replied:
This should be corrected in ITK :)

thewtex reviewed (approved):
:1st_place_medal:

fbudin69500 committed:
ENH: SimpleITK enums declared in an independent header

fbudin69500 committed:
BUG: Use filter_type value only when not empty

filter_type is used to define the type of the filter if it is not
the expected type (FilterName<InputImageType,OutputImageType>).
When parsing the JSON file and the dictionaries to verify the keys
available, filter_type is always initialized to an empty string and
therefore the key is always in the dictionary. We need to check
its value to know if we should use it or not instead of only checking
if the key is in the dictionary.

fbudin69500 committed:
ENH: Adding ITKWhiteTopHatImage filter

fbudin69500 committed:
ENH: WhiteTopHat Image filter uses an isotropic kernel

Kernel radius is set with one integer value instead of a vector.

fbudin69500 commented:
@thewtex Addition of the WhiteTopHat and a little debugging later, here is the updated PR.

thewtex commented on Documentation/ITKImageProcessingFilters/ITKWhiteTopHatImage.md:
```diff
@@ -0,0 +1,47 @@
+ITKImageProcessing (ITKImageProcessing)
+
+## Description ##
+White top hat extract local maxima that are larger than the structuring element.
```
extract -> extracts

fbudin69500 committed:
ENH: Adding all ITK Kernel Filters that were not already included

fbudin69500 commented:
@thewtex : I think I am done working on this branch :)

thewtex commented:
@fbudin69500 Wow! :-O Excellent!

One tweak -- GrayscaleGeodesicErodeImageFilter and GrayscaleGeodesicDilateImageFilter are deprecated and can be removed. This could be added in an Issue if not done on this branch.

imikejackson commented:
+1

fbudin69500 committed:
ENH: Remove deprecated ITKGrayscaleGeodesicDilateImage and ITKGrayscaleGeodesicErodeImage
